### PR TITLE
Extract evaluation procedure from training extension `Evaluator`

### PR DESCRIPTION
**Pull request by cocuh**
Add `chainer.training.evaluator.Evaluator` and `chainer.training.evaluator.StandardEvaluator` to evaluate testing error and accuracy.
Although `chainer.training.extensions.Evaluator` can evaluate them *by itself*, *standalone extension* is strange and it should be pure training extension which call external defined evaluation procedure. Thus, I define evaluator interface(`chainer.training.Evaluator`) and move evaluating codes from `chainer.training.extensions.Evaluator` to `chainer.training.evaluator.StandardEvaluator`.

I tested `examples/mnist/train_mnist.py` and it worked.

### CODE REVIEW

File: chainer/training/extensions/evaluator.py [hunk 1]
```diff
@@ -1,17 +1,10 @@
...
-
-import six
-
 from chainer.dataset import convert
-from chainer.dataset import iterator as iterator_module
-from chainer import link
 from chainer import reporter as reporter_module
 from chainer.training import extension
-from chainer import variable
+from chainer.training import StandardEvaluator
```
Comment (delta2323) [inline chainer/training/extensions/evaluator.py]: We forbid the direct import of class objects (See the [coding guideline](http://docs.chainer.org/en/stable/contribution.html#coding-guidelines) for detail). So could you replace it with `from chainer.training import evaluator` and use `evaluator.StandardEvaluator` instead?

File: chainer/training/evaluator.py [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+        >>>     (np.asarray([0, 1], dtype=np.float32), np.int32(1)),  # true
+        >>>     (np.asarray([0, 1], dtype=np.float32), np.int32(0)),  # false
+        >>> ]
+        >>>
+        >>> iterator = chainer.iterators.SerialIterator(data, 5, repeat=False)
+        >>>
+        >>> evaluator = StandardEvaluator(iterator, model)
+        >>> result = evaluator.run()
+        >>> print(result)
+        {'main/accuracy': 0.8000000119209289, 'main/loss': 0.5132616162300109}
```
Comment (delta2323) [inline chainer/training/evaluator.py]: I think this example is useful for testing `StandardEvaluator`. So, could you write unittests from it?

File: chainer/training/extensions/evaluator.py (hunk 1 above)
Comment (cocuh) [inline chainer/training/extensions/evaluator.py]: Oops I missed it.

**Comment by delta2323** [comment]
Thank you for updating. It looks good to me. I am now running tests and will merge it if tests will pass.

**Comment by cocuh** [comment]
Thank you for reviewing! This PR is bit large. Please decide carefully whether merge or not.

**Comment by delta2323** [comment]
LGTM 👍

**Comment by delta2323** [comment]
Thank you for your contribution!

**Comment by jekbradbury** [comment]
This breaks backwards compatibility in a slightly confusing way if you were using a custom Evaluator (now you have to pass that custom Evaluator to the new Evaluator extension as the `evaluator=` kwarg...but you still have to pass `iterator` and `target` positional arguments, even though they're completely ignored now).

**Comment by delta2323** [comment]
@jekbradbury Thank you for reporting the problem. I sent PR #1788 that hopefully fixes your problem. Could you check that to see if I understand your problem correctly?

**Comment by jekbradbury** [comment]
That helps if you are using Evaluator on its own (outside of Trainer) but the use case I'm talking about is when you are using an evaluator as an extension to Trainer. If you want to use a custom evaluator class as a Trainer extension, you now have to create that evaluator and pass it as a keyword argument `evaluator=` to the new Evaluator extension...but you also have to redundantly pass `iterator` and `target` as positional arguments even though they're ignored.

old: `trainer.extend(CustomEvaluator(dev_iter, model), eval_trigger)`
new: `trainer.extend(Evaluator(dev_iter, model, evaluator=CustomEvaluator(dev_iter, model)), eval_trigger)`

This change is documented, but it's still unintuitive. Maybe it would be better to give `iterator` and `target` defaults of `None`?

**Comment by delta2323** [comment]
@jekbradbury Thank you for the detail explanation. We found another problem on the backward compatibility. So, I think it is reasonable to revert this PR (#1790) and reopen this PR.

**Pull request merged**